examples/cosmos/ics20/ics20.qnt

```diff
@@ -0,0 +1,129 @@
+ *
+ * For now, this only models the denomination trace changes between transfers.
+ *
+ * Gabriela Moreira, Informal Systems, 2023
+ */
+module ics20 {
+  type Uint = int
+
+  type State = {
+    trace: List[str],
+    denom: str,
+  }
+
+  type FungibleTokenPacketData = {
+    trace: List[str],
+    denom: str,
+    amount: Uint,
+    sender: str,
+    receiver: str,
+  }
+
+  type Packet = {
+    data: FungibleTokenPacketData,
+    sourcePort: str,
+    sourceChannel: str,
+    destPort: str,
+    destChannel: str,
+  }
+
+  pure def onRecvPacket(packet: Packet): State =
+    pure val trace = packet.data.trace
+    pure val sourcePrefix = [packet.sourcePort, packet.sourceChannel]
+    if (trace.length() >= 2 and trace.slice(0, 2) == sourcePrefix) {
+      // Sender is source
+      // Remove prefix from trace
+      pure val newTrace = if (trace.length() == 2) [] else trace.slice(2, trace.length())
+      { trace: newTrace, denom: packet.data.denom }
+    } else {
+      // Sender is sink
+      // Add prefix to trace
+      pure val sinkPrefix = [packet.destPort, packet.destChannel]
+      { trace: sinkPrefix.concat(trace), denom: packet.data.denom }
+    }
+
+  var state: State
+
+  action init =
+    nondet denom = Set("denom1", "denom2", "denom3").oneOf()
+    state' = {
+      trace: [],
+      denom: denom,
+    }
+
+  // For testing, imagine the following channels in these chains setups
+  // PS: There should be similar channels between A and C, those are ommited in the diagram for simplicity
+  // ┌───────────────────┐           ┌────────────────────────────────────┐         ┌───────────────────┐
+  // │      Chain A      │           │               Chain B              │         │      Chain C      │
+  // │                   │           │                                    │         │                   │
+  // │ ┌───────────────┐ │           │ ┌──────────────┐  ┌──────────────┐ │         │ ┌───────────────┐ │
+  // │ │               │ │           │ │              │  │              │ │         │ │               │ │
+  // │ │  channelToB   │◄├───────────┤►│  channelToA  │  │  channelToC  │◄├─────────┤►│  channelToB   │ │
+  // │ │               │ │           │ │              │  │              │ │         │ │               │ │
+  // │ └───────────────┘ │           │ └──────────────┘  └──────────────┘ │         │ └───────────────┘ │
+  // │                   │           │                                    │         │                   │
+  // └───────────────────┘           └────────────────────────────────────┘         └───────────────────┘
+  pure def buildPacket(currentState: State, source: str, dest: str): Packet =
+    pure val channels = Map(
+      "A" -> "channelToA",
+      "B" -> "channelToB",
+      "C" -> "channelToC"
+    )
+    ({
+      data: {
+        trace: currentState.trace,
+        denom: currentState.denom,
+        amount: 100,
+        sender: source,
+        receiver: dest,
```

Comment:
We're not currently using these packet fields, but you're currently setting them to chain names. They're actually accounts.

Reply:
Ah, right!